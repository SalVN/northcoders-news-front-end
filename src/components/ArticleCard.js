import React from 'react';
import PropTypes from 'prop-types';
import { BrowserRouter as Router, Link } from 'react-router-dom';

const ArticleCard = function (props) {
  return (
    <div className='box'>
      <article className='media'>
        <div className='media-left'>
          {props.number + 1}
        </div>
        <div className='media-left'>
          <button className='button' onClick={props.voteHandler.bind(this, 'up', props.id)}><i className="fa fa-arrow-up" aria-hidden="true"></i></button>
          <div>{props.votes}</div>
          <button className='button' onClick={props.voteHandler.bind(this, 'down', props.id)}><i className="fa fa-arrow-down" aria-hidden="true"></i></button>
        </div>
        <div className='media-content'>
          <div className='content'>
            <Link to={`/articles/${props.id}`}>
              <h3 className='title is-3'>{props.title}</h3>
            </Link>
            <p>
              <span>added by: {props.author}</span>
              <span>tags: {props.tags}</span>
              <span>comments: {props.comment_count}</span>
            </p>
          </div>
        </div>
      </article>
    </div>
  );
};

ArticleCard.propTypes = {
  title: PropTypes.string.isRequired,
  votes: PropTypes.number.isRequired,
  number: PropTypes.number.isRequired,
  author: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired,
  tags: PropTypes.string.isRequired,
  comment_count: PropTypes.number.isRequired,
  voteHandler: PropTypes.func.isRequired
};


export default ArticleCard;
