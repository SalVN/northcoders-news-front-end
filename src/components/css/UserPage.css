.user-page-no-articles {
    margin-left: 20px;
    padding: 20px;
    font-size: 14px;
    color: black;
    font-weight: 400;
}