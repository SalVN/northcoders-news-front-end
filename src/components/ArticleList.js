import React from 'react';
import ArticleCard from './ArticleCard';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import * as actions from '../actions/actions';

class ArticleList extends React.Component {
  componentDidMount() {
    if (this.props.users && this.props.users.length < 1) {
      this.props.fetchUsers();
    }
  }
  render() {
    const users = this.props.users;
    const voteHandler = this.props.voteArticle;
    return (
      <div id='ArticleList' className='content'>
        {this.props.articles.length > 0 &&
          this.props.articles.map((article, i) => {
            const username = article.created_by;
            let index;
            users &&
              users.forEach((user, i) => {
                if (user.username === username) index = i;
              });
            return (
              <ArticleCard
                key={article._id}
                title={article.title}
                number={i}
                votes={article.votes}
                author={article.created_by}
                id={article._id}
                tags={article.belongs_to}
                comment_count={article.comment_count}
                voteHandler={voteHandler}
                userData={users[index]}
              />
            );
          })
        }
      </div>
    );
  }
}

function mapDispatchToProps(dispatch) {
  return {
    fetchUsers: () => {
      dispatch(actions.fetchUsers());
    }
  };
}

function mapStateToProps(state) {
  return {
    loading: state.users.loading,
    users: state.users.users
  };
}

ArticleList.propTypes = {
  articles: PropTypes.array.isRequired,
  loading: PropTypes.bool.isRequired,
  fetchUsers: PropTypes.func.isRequired,
  voteArticle: PropTypes.func.isRequired,
  users: PropTypes.array.isRequired
};

export default connect(mapStateToProps, mapDispatchToProps)(ArticleList);