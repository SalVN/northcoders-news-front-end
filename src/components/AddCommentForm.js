import React from 'react';
import PropTypes from 'prop-types';

import './css/AddCommentForm.css';

class AddCommentForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            text: ''
        };
        this.handleChange = this.handleChange.bind(this);
    }
    render() {
        return (
            <div>
                {this.props.showForm
                    ?
                    <div className='box box-style box-style-comment box-style-comment'>
                        <article className='media'>
                            <div className='media-left'>
                                <img className='comment-form-image' src={this.props.user.avatar_url} />
                            </div>
                            <div className='media-content'>
                                <div className='content'>
                                    <form onSubmit={this.props.handleSubmit}>
                                        <label className='comment-form-heading' htmlFor="comment-input">Join the conversation:</label>
                                        <br />
                                        <div className='comment-form' onBlur={this.props.toggleForm.bind(this, this.state.text)}>
                                            <textarea className='add-comment-textarea' onChange={this.handleChange} id="comment-input" defaultValue={this.state.text} placeholder='Join the conversation' />
                                        </div>
                                        <div className='button-div'>
                                            <button className='button add-comment-submit' type='submit'>Submit</button>
                                        </div>
                                    </form>
                                </div>
                            </div>
                        </article>
                    </div>
                    : <div className='box box-style box-style-comment box-style-comment-small'>
                        <article className='media'>
                            <div className='media-left'>
                                <img className='comment-form-image' src={this.props.user.avatar_url} />
                            </div>
                            <div className='media-content'>
                                <div className='content'>
                                    <div onClick={this.props.toggleForm.bind(this, null)} className='comment-form-small'>
                                        <p>Join the conversation</p>
                                    </div>
                                </div>
                            </div>
                        </article>
                    </div>
                }
            </div>
        );
    }
    handleChange(e) {
        this.setState({
            text: e.target.value
        });
    }
}

AddCommentForm.propTypes = {
    id: PropTypes.string.isRequired,
    addComment: PropTypes.func.isRequired,
    toggleForm: PropTypes.func.isRequired,
    user: PropTypes.object.isRequired,
    showForm: PropTypes.bool.isRequired,
    handleSubmit: PropTypes.func.isRequired
};

export default AddCommentForm;