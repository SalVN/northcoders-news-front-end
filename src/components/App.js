import React, {Component} from 'react';
import PropTypes from 'prop-types';

import TopicNavBar from './TopicNavBar';
import MainBanner from './MainBanner';
import UserCard from './UserCard';
import UserListCard from './UserListCard';

import './css/App.css';

export class App extends Component {
  render() {
    let pageStyling = window.innerWidth < 800 ? 'main-page-mobile' : 'main-page';
    return (
      <div className='page'>
        <TopicNavBar />
        <MainBanner />
        <div className={pageStyling}>
          <div className={'columns is-gapless'}>
            <div className='column is-three-quarters'>
              {this.props.children}
            </div>
            <div className='column is-one-quarter'>
              <div className='is-hidden-mobile'>
                <UserCard />
              </div>
              <UserListCard />
            </div>
          </div>
        </div>
      </div>
    );
  }
}

App.propTypes = {
  children: PropTypes.element.isRequired
};

export default App;
