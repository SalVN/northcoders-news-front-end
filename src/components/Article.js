import React from 'react';
import PropTypes from 'prop-types';
import { BrowserRouter as Router, Link } from 'react-router-dom';

import './css/Article.css';

const Article = function (props) {
    return (
        <div className='box'>
            <div className='columns'>
                <div className='column is-1'>
                    <button className='button is-black is-inverted is-small'><i className="fa fa-arrow-up" aria-hidden="true"></i></button>
                    <p className='comment-card-votes'><strong>{props.article.votes}</strong></p>
                    <button className='button is-black is-inverted is-small'><i className="fa fa-arrow-down" aria-hidden="true"></i></button>
                </div>
                <div className='column'>
                    <div className='title is-3 article-title'>{props.article.title}</div>
                    <p>
                        <span>
                            <span className='article-subheading-key'>added by:</span>
                            <Link to='#'>
                                <span>
                                    {props.user &&
                                        <img className='article-card-avatar' src={props.user.avatar_url} alt='user avatar' />
                                    }
                                </span>
                                <span className='article-subheading-value'>
                                    <strong>{props.article.created_by}</strong>
                                </span>
                            </Link>
                        </span>
                        <span>
                            <span className='article-card-subheading-key'>tags:</span>
                            <Link to={`/topics/${props.article.belongs_to}/articles`}>
                                <span className='article-card-subheading-value'>
                                    <strong>{props.article.belongs_to}</strong>
                                </span>
                            </Link>
                        </span>
                        <span>
                            <span className='article-card-subheading-key'>comments:</span>
                            {/*<Link to='#'>*/}
                            <span className='article-card-subheading-value-notlink'>
                                <strong>{props.article.commentCount}</strong>
                            </span>
                            {/*</Link>*/}
                        </span>
                    </p>
                    <div>{props.article.body}</div>
                </div>
            </div>
        </div>
    );
};

Article.propTypes = {
    article: PropTypes.object.isRequired,
    user: PropTypes.object.isRequired
};

export default Article;