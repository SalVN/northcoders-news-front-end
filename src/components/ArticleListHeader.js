import React from 'react';
import PropTypes from 'prop-types';

import './css/MainArticleList.css';

const ArticleListHeader = function (props) {
    return (
        <div className='content'>
            <div className='columns'>
                <div className='column is-3'>
                    <h2 className='page-title'><strong>{props.title}</strong></h2>
                </div>
                <div className='column is-1 is-offset-5 label'>
                    <label htmlFor="Sort by">Sort by:</label>
                </div>
                <div className='column is-3'>
                    <div className='select is-primary select-menu'>
                        <select name="Sort by" onChange={props.handleClickSelect}>
                            <option value="votes">Popularity</option>
                            <option value="comments">Number of Comments</option>
                            <option value="random">Random</option>
                        </select>
                    </div>
                </div>
            </div>
        </div>
    );
};

ArticleListHeader.propTypes = {
    title: PropTypes.string.isRequired,
    handleClickSelect: PropTypes.func.isRequired
};

export default ArticleListHeader;