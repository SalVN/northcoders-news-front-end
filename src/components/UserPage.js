import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import * as actions from '../actions/actions';
import { getIndexUsername } from '../utilities/getIndex';

import UserPageCard from './UserPageCard';
import { returnFilteredArticles } from '../utilities/filterArticles';
import './css/UserPage.css';

class UserPage extends Component {
    componentDidMount() {
        if (this.props.users && this.props.users.length < 1) {
            this.props.fetchUsers();
        }
        if (this.props.users && this.props.users.length < 1) {
            this.props.fetchArticles();
        }
    }
    render() {
        let index;
        let userArticles;
        if (this.props.users && this.props.users.length > 0) {
            this.props.users.sort((a, b) => {
                return (b.articles_vote_count + b.comments_vote_count) - (a.articles_vote_count + a.comments_vote_count);
            });
            index = getIndexUsername(this.props.users, this.props.match.params.id);
            userArticles = returnFilteredArticles(this.props.articles, this.props.users[index].username);
        }

        return (
            <div>
                {((this.props.users && this.props.users.length > 0) && (this.props.articles && this.props.articles.length > 0)) &&
                    <UserPageCard articlesNo={userArticles.length} ranking={index + 1} user={this.props.users[index]} />
                }
            </div>
        );
    }
}

function mapDispatchToProps(dispatch) {
    return {
        fetchUsers: () => {
            dispatch(actions.fetchUsers());
        },
        fetchArticles: () => {
            dispatch(actions.fetchArticles());
        }
    };
}

function mapStateToProps(state) {
    return {
        users: state.users.users,
        articles: state.articles.articles
    };
}


UserPage.propTypes = {
    users: PropTypes.array.isRequired,
    fetchUsers: PropTypes.func.isRequired,
    match: PropTypes.object.isRequired,
    fetchArticles: PropTypes.func.isRequired,
    articles: PropTypes.array.isRequired
};

export default connect(mapStateToProps, mapDispatchToProps)(UserPage);