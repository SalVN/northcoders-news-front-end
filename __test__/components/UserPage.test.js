import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { Route, MemoryRouter } from 'react-router-dom';

const mockStore = configureStore();
const initialState = {};

import { UserPage } from '../../src/components/UserPage';

describe('UserPage', () => {
    const users = [{
        _id: '59b1b18b327cce1fb043bdb1',
        username: 'northcoder',
        name: 'Awesome Northcoder',
        avatar_url: 'https://avatars3.githubusercontent.com/u/6791502?v=3&s=200',
        __v: 0
    }, {
        _id: '59b1b18b327cce1fb043bdb2',
        username: 'javascripter',
        name: 'Awesome Javascripter',
        avatar_url: 'https://avatars3.githubusercontent.com/u/6791502?v=3&s=200',
        __v: 0
    }];

    it('is a function', () => {
        expect(typeof UserPage).toEqual('function');
    });

    it('renders', () => {
        const store = mockStore(initialState);
        const enzymeWrapper = shallow(<UserPage
            users={users}
            match={
                {
                    params: {
                        id: '59b01acf006c8dbca914672f'
                    }
                }}
            fetchArticles={x => x}
            store={store}
        />);
        expect(enzymeWrapper.children().length).toEqual(4);
    });

    it('renders correctly', () => {
        const store = mockStore(initialState);
        const tree = renderer.create(
            <Provider>
                <MemoryRouter>
                    <UserPage
                        users={users}
                        match={
                            {
                                params: {
                                    id: '59b01acf006c8dbca914672f'
                                }
                            }}
                        fetchArticles={x => x}
                        store={store} />
                </MemoryRouter>
            </Provider>
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });
});