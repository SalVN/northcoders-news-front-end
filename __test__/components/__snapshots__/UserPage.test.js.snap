// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserPage renders correctly 1`] = `
<div>
  <div
    className="user-page-card-box"
  >
    <div
      className="container-fluid"
    >
      <div
        className="card"
      >
        <div
          className="user-page-card-background"
        >
          <div
            className="columns is-mobile"
          >
            <div
              className="column is-half-mobile is-one-quarter-tablet avatar-container"
            >
              <img
                className="user-card-avatar-page"
                src="https://avatars3.githubusercontent.com/u/6791502?v=3&s=200"
              />
            </div>
            <div
              className="column is-half-mobile is-three-quarters-tablet"
            >
              <p
                className="user-page-card-name"
              >
                Awesome Northcoder
              </p>
              <p
                className="user-page-card-username"
              >
                @northcoder
              </p>
            </div>
          </div>
        </div>
        <div
          className="card-content content-container"
        >
          <div
            className="columns is-mobile"
          >
            <div
              className="column is-2-tablet"
            >
              <div
                className="user-page-card-mini-title"
              >
                Popularity
              </div>
              <div
                className="user-page-card-number"
              >
                NaN
              </div>
            </div>
            <div
              className="column is-2-tablet"
            >
              <div
                className="user-page-card-mini-title"
              >
                Comments
              </div>
              <div
                className="user-page-card-number"
              />
            </div>
            <div
              className="column is-2-tablet"
            >
              <div
                className="user-page-card-mini-title"
              >
                Articles
              </div>
              <div
                className="user-page-card-number"
              >
                0
              </div>
            </div>
            <div
              className="column is-2-tablet is-hidden-mobile"
            >
              <div
                className="user-page-card-mini-title"
              >
                Comment Votes
              </div>
              <div
                className="user-page-card-number"
              />
            </div>
            <div
              className="column is-2-tablet is-hidden-mobile"
            >
              <div
                className="user-page-card-mini-title"
              >
                Article Votes
              </div>
              <div
                className="user-page-card-number"
              />
            </div>
            <div
              className="column is-2-tablet is-hidden-mobile"
            >
              <div
                className="user-page-card-mini-title"
              >
                Ranking
              </div>
              <div
                className="user-page-card-number"
              >
                1
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <hr />
  <div
    className="content"
    id="Articles"
  >
    <div
      className="content"
    >
      <div
        className="columns"
      >
        <div
          className="column"
        >
          <h2
            className="page-title"
          >
            <strong>
               Articles
            </strong>
          </h2>
        </div>
        <div
          className="column label has-text-left-mobile has-text-right-tablet"
        >
          <div
            className="dropdown has-text-left "
          >
            <div
              className="dropdown-trigger"
            >
              <button
                aria-controls="dropdown-menu"
                aria-haspopup="true"
                className="button dropdown-button-article-list"
                onClick={[Function]}
              >
                <label
                  htmlFor="Sort by"
                >
                  Sort by:
                </label>
                <span
                  className="icon"
                >
                  <i
                    aria-hidden="true"
                    className="fa fa-caret-down"
                  />
                </span>
              </button>
            </div>
            <div
              className="dropdown-menu"
              id="dropdown-menu"
              role="menu"
            >
              <div
                className="dropdown-content"
              >
                <a
                  className="dropdown-item popular"
                  onClick={[Function]}
                  value="votes"
                >
                  Popularity
                </a>
                <a
                  className="dropdown-item comments"
                  onClick={[Function]}
                  value="comments"
                >
                  Number of Comments
                </a>
                <a
                  className="dropdown-item random"
                  onClick={[Function]}
                  value="random"
                >
                  Random
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <p
      className="user-page-no-articles"
    >
      This user hasn't uploaded any articles yet
    </p>
  </div>
  <hr />
</div>
`;
